### PR TITLE
ethapi: estimate gas multiple consecutive calls on the same state

### PR DESCRIPTION
**Pull request by kvhnuke**
There are multiple occasions where you have to estimate the gas limit in a sequential order, meaning estimate should take state changes made by previous calls into consideration. Current `eth_estimateGas` creates a copy of the pending state on each call. Because of this, it is not possible to send out multiple txs that depends on each other at the same time.
Good example will be DEXs which requires tokens to be `approve(spender, amount)` before executing the trade. Currently some DEXs maintains a gas limit map since the gas limit cannot be estimated using `eth_estimateGas` without having the tokens approved.
At MEW we have to guess the gas limit for batch transactions where we will guess extreme amounts just to be on the safe side and this adds txs that are set with unnecessarily higher gas limits to the pending pool.

This PR will let users estimate gas on the same state. Also it is backwards compatible with current ethereum rpc standard. 
`{"jsonrpc":"2.0","method":"eth_estimateGasList","params":[[{call0-0}, {call0-1}, {call0-2}]],"id":1}`

```
{
  "id":1,
  "jsonrpc": "2.0",
  "result": ["0xvalue0-0", "0xvalue0-1", "0xvalue0-2"]
}
```
in this case `call0`s depends on previous call to be executed on the state.

backwards compatible
`{"jsonrpc":"2.0","method":"eth_estimateGas","params":[{call1-0}],"id":1}`

```
{
  "id":1,
  "jsonrpc": "2.0",
  "result": "0xvalue1-0"
}
```

### CODE REVIEW

File: internal/ethapi/api.go [hunk 1]
```diff
@@ -1000,24 +1006,50 @@ func DoEstimateGas(ctx context.Context, b Backend, args CallArgs, blockNrOrHash
...
+		validState = prevS
 	}
-	return hexutil.Uint64(hi), nil
+	return hexutil.Uint64(hi), validState, nil
 }
 
-// EstimateGas returns an estimate of the amount of gas needed to execute the
-// given transaction against the current pending block.
-func (s *PublicBlockChainAPI) EstimateGas(ctx context.Context, args CallArgs) (hexutil.Uint64, error) {
+func (s *PublicBlockChainAPI) EstimateGas(ctx context.Context, argsInterface interface{}) (interface{}, error) {
```
Comment (rjl493456442) [inline internal/ethapi/api.go:1023]: It's a little bit hacky to me. What about adding another API with `args []CallArgs` as the parameters?

Comment (kvhnuke) [reply internal/ethapi/api.go:1023]: Yea that works for me, how about eth_estimateGases ? Or do you have another one in mind. I didn't want to make changes to the standard API that's why just decided to go with the current implementation

Comment (rjl493456442) [reply internal/ethapi/api.go:1023]: Sounds good to me

Comment (kvhnuke) [reply internal/ethapi/api.go]: `estimateGases` sounds weird to me haha
I added `eth_estimateGasList`

**Comment by adamschmideg** [comment]
Closing in favor of #21268

**Pull request closed without merging**